simulator/memory/argv_loader/argv_loader.cpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * argv_loader.cpp: loader of arguments to guest memory
+ * @author Rustem Yunusov <yunusov.rr@phystech.edu>
+ * Copyright 2019 MIPT-MIPS iLab project
+ */
+
+#include "argv_loader.h"
+
+#include <cstring>
+#include <infra/exception.h>
+
+template <typename T, Endian endian>
+ArgvLoader<T, endian>::ArgvLoader( const char* const* argv, const char* const* envp)
+        : argc( argv ? count_argc( argv) : 0)
+        , argv( argv)
+        , envp( envp)
+        , offset( 0)
+{}
+
+template <typename T, Endian endian>
+size_t ArgvLoader<T, endian>::load_to( const std::shared_ptr<FuncMemory>& mem, Addr addr)
+{
+    mem->write<T, endian>( static_cast<T>( argc), addr + offset);
+    offset += bytewidth<T>;
+
+    offset += argc * bytewidth<Addr>; //reserved space for argv[]
```

Comment:
Note that you have to use `bytewidth<T>` here and everywhere.
I suggest to declare it as a constexpr variable in ArgvLoader:
```c++
static const constexpr size_t WORD_SIZE = bytewidth<T>;
```

Reply:
@BEANefiT one more place